docker/manager.py

```diff
@@ -0,0 +1,742 @@
+#!/usr/bin/env python3
+import itertools
+import json
+import operator
+import os
+import re
+import shutil
+from collections import defaultdict
+from copy import deepcopy
+from functools import lru_cache
+from pathlib import Path
+from typing import Dict, Iterator, List, MutableMapping, Optional, Tuple
+
+from absl import app, logging
+from dotenv import load_dotenv
+from jinja2 import Environment
+from requests import Session
+from utils import (
+    FLAGS,
+    ColoredFormatter,
+    cached_property,
+    flatten,
+    get_data,
+    get_nested,
+    load_manifest_yaml,
+    mapfunc,
+    maxkeys,
+    mkdir_p,
+    pprint,
+    set_data,
+    sprint,
+    walk,
+)
+
+from docker import DockerClient
+from docker.errors import APIError, BuildError, ImageNotFound
+
+README_TEMPLATE: Path = Path('./templates/docs/README.md.j2')
+
+DOCKERFILE_NAME: str = "Dockerfile"
+DOCKERFILE_TEMPLATE_SUFFIX: str = ".Dockerfile.j2"
+DOCKERFILE_BUILD_HIERARCHY: List[str] = ["base", "runtime", "cudnn", "devel"]
+DOCKERFILE_NVIDIA_REGEX: re.Pattern = re.compile(r'(?:nvidia|cuda|cudnn)+')
+
+SUPPORTED_PYTHON_VERSION: List[str] = ["3.6", "3.7", "3.8"]
+
+NVIDIA_REPO_URL: str = "https://developer.download.nvidia.com/compute/cuda/repos/{}/x86_64"
+NVIDIA_ML_REPO_URL: str = (
+    "https://developer.download.nvidia.com/compute/machine-learning/repos/{}/x86_64"
+)
+
+HTTP_RETRY_ATTEMPTS: int = 2
+HTTP_RETRY_WAIT_SECS: int = 20
+
+# setup some default
+logging.get_absl_handler().setFormatter(ColoredFormatter())
+log = logging.get_absl_logger()
+
+if not os.path.exists('./.env'):
+    log.warning(f"Make sure to create .env file at {os.getcwd()}")
+else:
+    load_dotenv()
+
+if os.geteuid() == 0:
+    # We only use docker_client when running as root.
+    log.debug("Creating an instance of DockerClient")
+    docker_client = DockerClient(
+        base_url="unix://var/run/docker.sock", timeout=3600, tls=True
+    )
+
+
+class LogsMixin(object):
+    """Logs Mixin for docker build process"""
+
+    def build_logs(self, resp: Iterator, image_tag: str):
```

Comment:
Your documentation overall is very good. I would appreciate some more explicit, extended comments associated with your function headers, generally, similar to what you do with some of your classes, like so:
```suggestion
    """
    does what build_logs() does, using what it uses, etc.
    Args:
        resp: is what a resp is, does what a resp does
        image_tag: is what an image_tag is, does what an image_tag does
    """
    def build_logs(self, resp: Iterator, image_tag: str):
```